common/djangoapps/student/models.py

```diff
@@ -3116,6 +3122,135 @@ class AccountRecoveryConfiguration(ConfigurationModel):
+                 self.last_day_of_streak, self.last_streak_celebration, self.longest_ever_streak)
+
+    @classmethod
+    def _get_now(cls):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    @classmethod
+    def _calculate_streak_updates(cls, celebration, now):
+        """ Calculate the updates that should me applied to the streak fields of the provided celebration """
+        first_day_of_streak = celebration.first_day_of_streak
+        last_day_of_streak = celebration.last_day_of_streak
+        last_streak_celebration = celebration.last_streak_celebration
+        today = now.date()
+        yesterday = today - timedelta(days=1)
+        should_celebrate = False
+
+        first_ever_streak = None is last_streak_celebration is first_day_of_streak is last_day_of_streak
+        in_active_streak = last_day_of_streak and first_day_of_streak
+        broken_streak = last_day_of_streak not in (yesterday, today)
+        can_start_new_streak = last_day_of_streak and last_day_of_streak + timedelta(days=STREAK_BREAK_LENGTH) < today
+
+        if first_ever_streak:
+            # Start new streak
+            first_day_of_streak = today
+            last_day_of_streak = today
+        elif broken_streak:
+            if can_start_new_streak:
+                # If we were on a break and the break is over start new streak
+                first_day_of_streak = today
+                last_day_of_streak = today
+        elif in_active_streak:
+            # Increment existing streak
+            last_day_of_streak = today
+
+            streak_length = (last_day_of_streak - first_day_of_streak).days + 1
+            if streak_length == STREAK_LENGTH_TO_CELEBRATE:
+                # Celebrate if we didn't already celebrate today
+                if not last_streak_celebration == today:
+                    last_streak_celebration = today
+                    should_celebrate = True
+
+        return (first_day_of_streak, last_day_of_streak, last_streak_celebration), should_celebrate
+
+    @classmethod
+    def _update_streak(cls, celebration, streak_dates):
+        """ Update the celebration with the new streak data """
+        (first_day_of_streak, last_day_of_streak, last_streak_celebration) = streak_dates
+        # If anything needs to be updated, update the celebration in the database
+        if (first_day_of_streak != celebration.first_day_of_streak or
+                last_day_of_streak != celebration.last_day_of_streak or
+                last_streak_celebration != celebration.last_streak_celebration):
```

Comment:
I really really wish Django had a "did any fields change" method. We had to build a mixin that provided something like that in Disco. But it felt a little gross, because it relied on caching copies of the values.

That said, `_calculate_streak_updates` should be able to already know whether it changed anything. Couldn't it just return `did_change, should_celebrate` rather than all the individual values?

Reply:
I was suggesting less return values to `calculate_streak_updates`, not more. Instead of returning 3 (`first_day_of_streak, last_day_of_streak, streak_length_to_celebrate`), I was suggesting just 2 (`changed_something, streak_length_to_celebrate`).

Just because `calculate_streak` knows if anything changed, so that's easy to return vs rechecking each field again (I think the logic might be as simple as `not already_updated_streak_today`?)

Reply:
OK, after talking about this on slack, I retract this thread - I like the way you've done it and my idea is worse. :+1: